### PR TITLE
feat(FTL-18291): implementa credential revocation

### PR DESCRIPTION
**Pull request by YoussefAWasfy**

### CODE REVIEW

**Review by linn-affinidi — DISMISSED** [review]
LGTM

File: generator/nextjs/template/src/components/issuance/Offer.tsx [hunk 1]
```diff
@@ -3,8 +3,28 @@ import QRCode from "qrcode.react";
...
+    if (vaultUrl) {
+      return buildClaimLinkInternal(
+        vaultUrl,
+        credentialOfferUri
+      )
+    }
+  }
+  return VaultUtils.buildClaimLink(credentialOfferUri)
+}
+
```
Comment (carlos-affinidi) [inline generator/nextjs/template/src/components/issuance/Offer.tsx:25]: These and the `apiGatewayUrl` are internal changes to run the app on dev environment. Please don't merge them to main as they are not meant to be used by developers. Either remove change the MR target to the `with-env` branch.

File: samples/affinidi-nextjs-nextauthjs/src/components/NavBar.tsx [hunk 2]
```diff
@@ -1,9 +1,9 @@
...
-import Link from "next/link";
-import { FC, useEffect, useState } from "react";
-import Button from "./core/Button";
-import { clientLogin } from "src/lib/auth/client-login";
+import { Session } from 'next-auth';
+import { getSession, signOut } from 'next-auth/react';
+import Link from 'next/link';
+import { FC, useEffect, useState } from 'react';
+import Button from './core/Button';
+import { clientLogin } from 'src/lib/auth/client-login';
```
Comment (carlos-affinidi) [inline samples/affinidi-nextjs-nextauthjs/src/components/NavBar.tsx:6]: Ref app uses the default prettier configuration, not the internal one.

File: samples/affinidi-nextjs-nextauthjs/src/components/NavBar.tsx [hunk 3]
```diff
@@ -59,6 +59,13 @@ const NavBar: FC = () => {
           >
             Issue Credentials
           </Link>
+          <Link
+            href="/credential-revocation"
+            className="py-4 font-medium transition-colors hover:text-blue-500"
+            prefetch={false}
+          >
+            Revoke Credentials
+          </Link>
```
Comment (carlos-affinidi) [inline samples/affinidi-nextjs-nextauthjs/src/components/NavBar.tsx:68]: I see that multiple changes are done directly on the `samples/affinidi-nextjs-nextauthjs` directory. This is an autogenerated folder. Please make sure to add all of your changes to the `generator/nextjs/template` directory and run `npm run generate:nextjs` from the root. Working directly on the sample apps is usually a bad idea as you risk loosing your changes if you run the generator. See [CONTRIBUTING.md](https://github.com/affinidi/reference-app-affinidi-vault/blob/main/CONTRIBUTING.md) for more info.

File: samples/affinidi-nextjs-nextauthjs/src/components/issuance/CredentialsRevocationTable.tsx [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    const date = new Date(dateTime);
+    return date.toLocaleString('en-US', {
+      weekday: 'long', // e.g., "Wednesday"
+      year: 'numeric', // e.g., "2024"
+      month: 'long', // e.g., "December"
+      day: 'numeric', // e.g., "18"
+      hour: 'numeric', // e.g., "11 AM"
+      minute: '2-digit', // e.g., "23"
+    });
+  };
```
Comment (carlos-affinidi) [inline samples/affinidi-nextjs-nextauthjs/src/components/issuance/CredentialsRevocationTable.tsx:43]: Do we need a user friendly date? This app is meant for developers, and they may find a simple `date.toISOString()` more useful as it includes timezone.

File: samples/affinidi-nextjs-nextauthjs/src/pages/credential-revocation.tsx [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+const fetchIssuanceDataRecords = async (
+  issuanceConfigurationId: string,
+  exclusiveStartKey?: string
+): Promise<ListIssuanceRecordResponse> => {
+  const response = await fetch(
+    '/api/issuance/list-issuance-data-records?' +
+      new URLSearchParams({ issuanceConfigurationId }) +
+      '&' +
+      new URLSearchParams({ exclusiveStartKey: exclusiveStartKey! }),
```
Comment (carlos-affinidi) [inline samples/affinidi-nextjs-nextauthjs/src/pages/credential-revocation.tsx:21]: Suggestion to create a single `URLSearchParams` and call its `.toString()`

File: samples/affinidi-nextjs-nextauthjs/src/pages/credential-revocation.tsx [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+export const getServerSideProps = (async () => {
+  return { props: { featureAvailable: personalAccessTokenConfigured() } };
+}) satisfies GetServerSideProps<{ featureAvailable: boolean }>;
+
+export default function CredentialRevocation({
+  featureAvailable,
+}: InferGetServerSidePropsType<typeof getServerSideProps>) {
+  // Prefill did from session
+  // const [items, setItems] = useState<FlowData[]>();
```
Comment (carlos-affinidi) [inline samples/affinidi-nextjs-nextauthjs/src/pages/credential-revocation.tsx:64]:
```suggestion
```

**Review by carlos-affinidi — CHANGES_REQUESTED** [review]

File: samples/affinidi-nextjs-nextauthjs/src/components/issuance/CredentialsRevocationTable.tsx (hunk 4 above)
Comment (YoussefAWasfy) [reply samples/affinidi-nextjs-nextauthjs/src/components/issuance/CredentialsRevocationTable.tsx:43]: was a request from Digvijay

**Comment by carlos-affinidi** [comment]
@YoussefAWasfy One last request, as you are calling new endpoints please add the policy operation to [apps.json](https://github.com/affinidi/reference-app-affinidi-vault/blob/main/samples/apps.json#L355). This is used by CLI when generating the PAT for the ref app.

**Review by carlos-affinidi — APPROVED** [review]

**Review by carlos-affinidi — CHANGES_REQUESTED** [review]
Please re add the required policy permissions in `samples/apps.json`

File: samples/auth0-nextjs-nextauthjs/src/pages/credential-revocation.tsx [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    if (!session || !session.userId) {
+      return (
+        <div>
+          You must be logged in to issue credentials to your Affinidi Vault
+        </div>
+      );
+    }
+  };
```
Comment (aeffinidi) [inline samples/auth0-nextjs-nextauthjs/src/pages/credential-revocation.tsx:126]: embedded component is said to be a bad practice. It could be extracted out.

File: samples/auth0-nextjs-nextauthjs/src/components/issuance/CredentialsRevocationTable.tsx [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+        console.error("Error revoking:", error);
+        setHasError((error as any).message);
+      } finally {
+        if (hasError?.length === 0) setIsLoading(false); // Stop loading
+        setSelectedReason("");
+      }
+    }
+  };
+
+  const renderModal = () => {
```
Comment (aeffinidi) [inline samples/auth0-nextjs-nextauthjs/src/components/issuance/CredentialsRevocationTable.tsx:66]: same here, extract the internal component.

File: samples/affinidi-nextjs-nextauthjs/src/pages/credential-revocation.tsx [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    if (!session || !session.userId) {
+      return (
+        <div>
+          You must be logged in to issue credentials to your Affinidi Vault
+        </div>
+      );
+    }
+  };
```
Comment (aeffinidi) [inline samples/affinidi-nextjs-nextauthjs/src/pages/credential-revocation.tsx:126]: same here, but maybe you want to do that in another PR as this one is already taking long to be merged.

File: generator/nextjs/template/src/pages/iota-callback/index.tsx [hunk 10]
```diff
@@ -2,8 +2,15 @@ import { GetStaticProps } from "next";
 import { personalAccessTokenConfigured } from "src/lib/env";
 import { useSearchParams } from "next/navigation";
 import IotaCallbackPageError from "./iota-callback-error";
-import IotaCallbackContent from "./iota-callback-content";
 import { useIsClient } from "@uidotdev/usehooks";
```
Comment (aeffinidi) [inline generator/nextjs/template/src/pages/iota-callback/index.tsx:5]: the hook is not used anymore.

**Review by carlos-affinidi — APPROVED** [review]

File: samples/affinidi-nextjs-nextauthjs/src/pages/credential-revocation.tsx [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+  );
+
+  if (response.status === 200) {
+    const res = await response.json();
+    flowDataRecordsQuery.refetch();
+    return res;
+  } else {
+    throw Error(await response.text());
+  }
+};
```
[the remark this answers is not in the document]
Comment (aeffinidi) [reply samples/affinidi-nextjs-nextauthjs/src/pages/credential-revocation.tsx:50]: let's see whether it makes sens to do that in another PR , thank you

**Review by aeffinidi — APPROVED** [review]
LGTM!
